File: mParticle-Apple-SDK/Utils/MPUploadSettings.m
#import "MPUploadSettings.h"
#import "MPStateMachine.h"
#import "mParticle.h"
#import "MPILogger.h"

@import mParticle_Apple_SDK_Swift;

@interface MPUploadSettings ()
@property (nonatomic, strong, nonnull) MPUploadSettingsPRIVATE *impl;
@end

@implementation MPUploadSettings

+ (void)initialize {
    if (self == [MPUploadSettings class]) {
        [NSKeyedUnarchiver setClass:[MPUploadSettings class]
                       forClassName:@"mParticle_Apple_SDK.MPUploadSettings"];
        [NSKeyedUnarchiver setClass:[MPUploadSettings class]
                       forClassName:@"mParticle_Apple_SDK_NoLocation.MPUploadSettings"];
    }
}

+ (BOOL)supportsSecureCoding {
    return YES;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _impl = [[MPUploadSettingsPRIVATE alloc] init];
    }
    return self;
}

- (NSString *)apiKey { return self.impl.apiKey; }
- (void)setApiKey:(NSString *)apiKey { self.impl.apiKey = apiKey; }
- (NSString *)secret { return self.impl.secret; }
- (void)setSecret:(NSString *)secret { self.impl.secret = secret; }
- (NSString *)eventsHost { return self.impl.eventsHost; }
- (void)setEventsHost:(NSString *)eventsHost { self.impl.eventsHost = eventsHost; }
- (NSString *)eventsTrackingHost { return self.impl.eventsTrackingHost; }
- (void)setEventsTrackingHost:(NSString *)eventsTrackingHost { self.impl.eventsTrackingHost = eventsTrackingHost; }
- (BOOL)overridesEventsSubdirectory { return self.impl.overridesEventsSubdirectory; }
- (void)setOverridesEventsSubdirectory:(BOOL)overridesEventsSubdirectory { self.impl.overridesEventsSubdirectory = overridesEventsSubdirectory; }
- (NSString *)aliasHost { return self.impl.aliasHost; }
- (void)setAliasHost:(NSString *)aliasHost { self.impl.aliasHost = aliasHost; }
- (NSString *)aliasTrackingHost { return self.impl.aliasTrackingHost; }
- (void)setAliasTrackingHost:(NSString *)aliasTrackingHost { self.impl.aliasTrackingHost = aliasTrackingHost; }
- (BOOL)overridesAliasSubdirectory { return self.impl.overridesAliasSubdirectory; }
- (void)setOverridesAliasSubdirectory:(BOOL)overridesAliasSubdirectory { self.impl.overridesAliasSubdirectory = overridesAliasSubdirectory; }
- (BOOL)eventsOnly { return self.impl.eventsOnly; }
- (void)setEventsOnly:(BOOL)eventsOnly { self.impl.eventsOnly = eventsOnly; }

- (id)copyWithZone:(NSZone *)zone {
    MPUploadSettings *copy = [[MPUploadSettings alloc] initWithApiKey:self.apiKey
                                                                  secret:self.secret
                                                             eventsHost:self.eventsHost
                                                      eventsTrackingHost:self.eventsTrackingHost
                                       overridesEventsSubdirectory:self.overridesEventsSubdirectory
                                                          aliasHost:self.aliasHost
                                                   aliasTrackingHost:self.aliasTrackingHost
                                        overridesAliasSubdirectory:self.overridesAliasSubdirectory
                                                        eventsOnly:self.eventsOnly];
    return copy;
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [self.impl encodeToCoder:coder];
}

- (nullable instancetype)initWithCoder:(NSCoder *)coder {
    self = [super init];
    if (self) {
        _impl = [[MPUploadSettingsPRIVATE alloc] initFromCoder:coder];
    }
    return self;
}

+ (nonnull instancetype)currentUploadSettingsWithStateMachine:(nonnull id<MPStateMachineProtocol>)stateMachine
                                                networkOptions:(nonnull MPNetworkOptions *)networkOptions {
    return [[MPUploadSettings alloc] initWithApiKey:stateMachine.apiKey
                                             secret:stateMachine.secret
                                     networkOptions:networkOptions];
}

- (nonnull instancetype)initWithApiKey:(nonnull NSString *)apiKey
                                 secret:(nonnull NSString *)secret
                         networkOptions:(nonnull MPNetworkOptions *)networkOptions {
    NSString *customHost = networkOptions.customBaseURL ? networkOptions.customBaseURL.host : nil;
    if (customHost) {
        if (networkOptions.eventsHost) {
            MPILogWarning(@"MPNetworkOptions: customBaseURL is set; eventsHost is ignored.");
        }
        if (networkOptions.eventsTrackingHost) {
            MPILogWarning(@"MPNetworkOptions: customBaseURL is set; eventsTrackingHost is ignored.");
        }
        if (networkOptions.aliasHost) {
            MPILogWarning(@"MPNetworkOptions: customBaseURL is set; aliasHost is ignored.");
        }
        if (networkOptions.aliasTrackingHost) {
            MPILogWarning(@"MPNetworkOptions: customBaseURL is set; aliasTrackingHost is ignored.");
        }
    }
    return [self initWithApiKey:apiKey
                         secret:secret
                    eventsHost:[MPUploadSettingsPRIVATE resolvedHostWithCustomHost:customHost host:networkOptions.eventsHost]
             eventsTrackingHost:[MPUploadSettingsPRIVATE resolvedHostWithCustomHost:customHost host:networkOptions.eventsTrackingHost]
  overridesEventsSubdirectory:networkOptions.overridesEventsSubdirectory
                     aliasHost:[MPUploadSettingsPRIVATE resolvedHostWithCustomHost:customHost host:networkOptions.aliasHost]
              aliasTrackingHost:[MPUploadSettingsPRIVATE resolvedHostWithCustomHost:customHost host:networkOptions.aliasTrackingHost]
   overridesAliasSubdirectory:networkOptions.overridesAliasSubdirectory
                   eventsOnly:networkOptions.eventsOnly];
}

- (nonnull instancetype)initWithApiKey:(nonnull NSString *)apiKey
                                 secret:(nonnull NSString *)secret
                            eventsHost:(nullable NSString *)eventsHost
                     eventsTrackingHost:(nullable NSString *)eventsTrackingHost
          overridesEventsSubdirectory:(BOOL)overridesEventsSubdirectory
                             aliasHost:(nullable NSString *)aliasHost
                      aliasTrackingHost:(nullable NSString *)aliasTrackingHost
           overridesAliasSubdirectory:(BOOL)overridesAliasSubdirectory
                           eventsOnly:(BOOL)eventsOnly {
    self = [super init];
    if (self) {
        _impl = [[MPUploadSettingsPRIVATE alloc] initWithApiKey:apiKey
                                                         secret:secret
                                                     eventsHost:eventsHost
                                             eventsTrackingHost:eventsTrackingHost
                                    overridesEventsSubdirectory:overridesEventsSubdirectory
                                                      aliasHost:aliasHost
                                              aliasTrackingHost:aliasTrackingHost
                                     overridesAliasSubdirectory:overridesAliasSubdirectory
                                                     eventsOnly:eventsOnly];
    }
    return self;
}

@end

File: mParticle-Apple-SDK/Utils/MPUploadSettings.swift
//
//  MPUploadSettings.swift
//  mParticle-Apple-SDK
//
//  Created by Brandon Stalnaker on 1/22/25.
//
private let kApiKey = "apiKey"
private let kSecret = "secret"
private let kEventsHost = "eventsHost"
private let kEventsTrackingHost = "eventsTrackingHost"
private let kOverridesEventsSubdirectory: String = "overridesEventsSubdirectory"
private let kAliasHost = "aliasHost"
private let kAliasTrackingHost = "aliasTrackingHost"
private let kOverridesAliasSubdirectory = "overridesAliasSubdirectory"
private let kEventsOnly = "eventsOnly"

@objc(MPUploadSettings)
public class MPUploadSettings: NSObject, NSCopying, NSSecureCoding {
    @objc public var apiKey: String
    @objc public var secret: String
    @objc public var eventsHost: String?
    @objc public var eventsTrackingHost: String?
    @objc public var overridesEventsSubdirectory: Bool = false
    @objc public var aliasHost: String?
    @objc public var aliasTrackingHost: String?
    @objc public var overridesAliasSubdirectory: Bool = false
    @objc public var eventsOnly: Bool = false

    @objc override public init() {
        apiKey = ""
        secret = ""
        super.init()
    }

    public func copy(with _: NSZone? = nil) -> Any {
        return MPUploadSettings(apiKey: apiKey,
                                secret: secret,
                                eventsHost: eventsHost,
                                eventsTrackingHost: eventsTrackingHost,
                                overridesEventsSubdirectory: overridesEventsSubdirectory,
                                aliasHost: aliasHost,
                                aliasTrackingHost: aliasTrackingHost,
                                overridesAliasSubdirectory: overridesAliasSubdirectory,
                                eventsOnly: eventsOnly)
    }

    public static var supportsSecureCoding: Bool = true

    public func encode(with coder: NSCoder) {
        coder.encode(apiKey, forKey: kApiKey)
        coder.encode(secret, forKey: kSecret)
        coder.encode(eventsHost, forKey: kEventsHost)
        coder.encode(eventsTrackingHost, forKey: kEventsTrackingHost)
        coder.encode(overridesEventsSubdirectory, forKey: kOverridesEventsSubdirectory)
        coder.encode(aliasHost, forKey: kAliasHost)
        coder.encode(aliasTrackingHost, forKey: kAliasTrackingHost)
        coder.encode(overridesAliasSubdirectory, forKey: kOverridesAliasSubdirectory)
        coder.encode(eventsOnly, forKey: kEventsOnly)
    }

    @objc public required init?(coder: NSCoder) {
        apiKey = coder.decodeObject(forKey: kApiKey) as? String ?? ""
        secret = coder.decodeObject(forKey: kSecret) as? String ?? ""
        eventsHost = coder.decodeObject(forKey: kEventsHost) as? String
        eventsTrackingHost = coder.decodeObject(forKey: kEventsTrackingHost) as? String
        overridesEventsSubdirectory = coder.decodeBool(forKey: kOverridesEventsSubdirectory)
        aliasHost = coder.decodeObject(forKey: kAliasHost) as? String
        aliasTrackingHost = coder.decodeObject(forKey: kAliasTrackingHost) as? String
        overridesAliasSubdirectory = coder.decodeBool(forKey: kOverridesAliasSubdirectory)
        eventsOnly = coder.decodeBool(forKey: kEventsOnly)
    }

    @objc public class func currentUploadSettings(stateMachine: MPStateMachineProtocol, networkOptions: MPNetworkOptions) -> MPUploadSettings {
        return MPUploadSettings(apiKey: stateMachine.apiKey, secret: stateMachine.secret, networkOptions: networkOptions)
    }

    @objc public init(apiKey: String, secret: String, eventsHost: String? = nil, eventsTrackingHost: String? = nil, overridesEventsSubdirectory: Bool = false, aliasHost: String? = nil, aliasTrackingHost: String? = nil, overridesAliasSubdirectory: Bool = false, eventsOnly: Bool = false) {
        self.apiKey = apiKey
        self.secret = secret
        self.eventsHost = eventsHost
        self.eventsTrackingHost = eventsTrackingHost
        self.overridesEventsSubdirectory = overridesEventsSubdirectory
        self.aliasHost = aliasHost
        self.aliasTrackingHost = aliasTrackingHost
        self.overridesAliasSubdirectory = overridesAliasSubdirectory
        self.eventsOnly = eventsOnly

        super.init()
    }

    @objc public init(apiKey: String, secret: String, networkOptions: MPNetworkOptions) {
        self.apiKey = apiKey
        self.secret = secret
        eventsHost = networkOptions.eventsHost
        eventsTrackingHost = networkOptions.eventsTrackingHost
        overridesEventsSubdirectory = networkOptions.overridesEventsSubdirectory
        aliasHost = networkOptions.aliasHost
        aliasTrackingHost = networkOptions.aliasTrackingHost
        overridesAliasSubdirectory = networkOptions.overridesAliasSubdirectory
        eventsOnly = networkOptions.eventsOnly

        super.init()
    }
}
